### PR TITLE
Nested class, interface or enum shouldn't be publicly visible

### PR DESCRIPTION
**Pull request by tagantroy**
Hello, Implementation for #444
Hacktoberfest!

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		if (isInternal(klass)) {
+			klass.declarations.forEach {
+				it.accept(visitor)
+			}
+		}
+	}
+
+	private fun isInternal(klass: KtClass): Boolean {
+		return klass.visibilityModifierType() == KtTokens.INTERNAL_KEYWORD
+	}
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: Move this to `Junk.kt` right under the following function.
```kotlin
fun KtModifierListOwner.isPublic()
```

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+	private fun isPrivate(klass: KtClass): Boolean {
+		return klass.visibilityModifierType() == KtTokens.PRIVATE_KEYWORD
+	}
+
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: Same here

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	private fun handleClass(klass: KtClass) {
+		report(CodeSmell(issue, Entity.from(klass)))
+	}
+
+	private class VisibilityVisitor(val rule: NestedClassesVisibility) : DetektVisitor() {
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: You can also use an inner class here.

Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: yes, thx!

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt (hunk 1 above)
Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: done

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt (hunk 2 above)
Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: done

**Comment by tagantroy** [comment]
@schalkms hello, can you review again?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/Junk.kt [hunk 4]
```diff
@@ -23,8 +23,12 @@ fun KtModifierListOwner.isPublicNotOverridden() =
 fun KtModifierListOwner.isPublic(): Boolean {
 	return this.hasModifier(KtTokens.PUBLIC_KEYWORD)
 			|| !(this.hasModifier(KtTokens.PRIVATE_KEYWORD)
-				|| this.hasModifier(KtTokens.PROTECTED_KEYWORD)
-				|| this.hasModifier(KtTokens.INTERNAL_KEYWORD))
+			|| this.hasModifier(KtTokens.PROTECTED_KEYWORD)
+			|| this.hasModifier(KtTokens.INTERNAL_KEYWORD))
+}
+
+fun KtModifierListOwner.isInternal(): Boolean {
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/Junk.kt]: let's do this?

```kotlin
fun KtModifierListOwner.isInternal() =
  hasModifier(KtTokens.INTERNAL_KEYWORD)
```

Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/Junk.kt]: @vanniktech Thx! done

File: detekt-rules/src/test/resources/cases/NestedClassesVisibility.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	public class NestedType { // report
+	}
+
+	enum class NestedEnum { One, Two } // report
+
+	private enum class NestedPrivateEnum { Three } // valid
+
+	interface Test {} // report
+
+	private interface PrivateTest {} // valid
```
Comment (arturbosch) [inline detekt-rules/src/test/resources/cases/NestedClassesVisibility.kt]: Please add some cases or nested classes (nest-level > 2)

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	override fun visitClass(klass: KtClass) {
+		super.visitClass(klass)
+		if (klass.isInternal()) {
+			klass.declarations.forEach {
+				it.accept(visitor)
+			}
+		}
+	}
+
+	private fun handleClass(klass: KtClass) {
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: This method could be inlined.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+		report(CodeSmell(issue, Entity.from(klass)))
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: Should we report the issue with a concrete description? Something along the lines of: `issue.copy(description = "Nested types are often used for implementing private functionality. However the visiblilty of ${klass.name} makes it visible externally.")

That mentions the class name and might make the description of the user a bit clearer to the user.

File: detekt-rules/src/test/resources/cases/NestedClassesVisibility.kt (hunk 5 above)
Comment (tagantroy) [reply detekt-rules/src/test/resources/cases/NestedClassesVisibility.kt]: done + fix for enums

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt (hunk 7 above)
Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NestedClassesVisibility.kt]: I agree. Done

**Comment by arturbosch** [comment]
Ok cool, thanks for the changes and contribution. I will wait for @schalkms to approve this now too and merge it :).

**Pull request merged**